src/util/network_namespace.cc

```diff
@@ -0,0 +1,168 @@
+/* -*-mode:c++; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*- */
+#include "network_namespace.hh"
+
+#include <stdio.h>
+#include <string.h>
+#include <unistd.h>
+#include <fcntl.h>
+#include <dirent.h>
+#include <errno.h>
+#include <sys/mount.h>
+#include <sys/stat.h>
+#include <sched.h>
+
+#include "config.h"
+#include "system_runner.hh"
+#include "exception.hh"
+#include "file_descriptor.hh"
+
+using namespace std;
+
+NetworkNamespace::NetworkNamespace( const std::string & name )
+    : name_(name), has_own_resolvconf_(false), has_entered_(false)
+{
+    /* create a new network namespace */
+    run( { IP, "netns", "add", name_ } );
+}
+
+
+NetworkNamespace::~NetworkNamespace()
+{
+    if ( has_entered_ ) {
+
+        char etc_netns_path[PATH_MAX];
+        char netns_name[PATH_MAX];
+        char etc_name[PATH_MAX];
+        struct dirent *entry;
+        DIR *dir;
+
+        snprintf( etc_netns_path, sizeof( etc_netns_path ), "%s/%s", NETNS_ETC_DIR.c_str(), name_.c_str() );
+        dir = opendir( etc_netns_path );
+        if (!dir )
+            return;
+
+        while ( ( entry = readdir(dir) ) != NULL ) {
+
+            if ( strcmp(entry->d_name, ".") == 0 )
+                continue;
+            if ( strcmp(entry->d_name, "..") == 0 )
+                continue;
+
+            snprintf(netns_name, sizeof(netns_name), "%s/%s", etc_netns_path, entry->d_name);
+            snprintf(etc_name, sizeof(etc_name), "/etc/%s", entry->d_name);
+
+            if ( umount( etc_name ) < 0 ) {
+                throw runtime_error( string("Unmounting... ") + netns_name + " -> " + etc_name + " failed: " + strerror(errno) + "\n" );
+            }
+        }
+
+    }
+
+    if ( has_own_resolvconf_ ) {
+        const string netns_dir = NETNS_ETC_DIR + "/" + name_;
+        const string resolvconf_path = netns_dir + "/resolv.conf";
+
+        SystemCall("unlink", unlink ( resolvconf_path.c_str() ) );
+        SystemCall("unlink", rmdir ( netns_dir.c_str() ) );
+    }
+
+    /*char netns_path[PATH_MAX];
+    snprintf(netns_path, sizeof(netns_path), "%s/%s", NETNS_DIR.c_str(), name_.c_str());
+    umount2(netns_path, MNT_FORCE);
+    if (unlink(netns_path) < 0) {
+      throw runtime_error(string("Cannot remove namespace file manually \"")+netns_path+"\":" + strerror(errno) + "\n" );
+    }*/
+
+
+    run( { IP, "netns", "delete", name_ } );
+}
+
+
+void NetworkNamespace::create_resolvconf( const std::string & nameserver )
+{
+  has_own_resolvconf_ = true;
+  /* create an individual resolv.conf for this network namespace */
+  const string netns_dir = NETNS_ETC_DIR + "/" + name_;
+  const string resolvconf_path = netns_dir + "/resolv.conf";
+
+  SystemCall("mkdir", mkdir( netns_dir.c_str() , S_IRWXU | S_IRWXG | S_IROTH | S_IXOTH) );
```

Comment:
This fails for me on a default Ubuntu 16.10 installation (there's no NETNS_ETC_DIR, aka `/etc/netns`, so the `mkdir` fails). We might need more like a mkdir -p, except how do you safely clean that up at the end of the program?